### ISSUE TITLE
TryCatchUpWithPrimary in a thread can not catch up primary

### ISSUE THREAD

**Issue by rockeet** [facebook/rocksdb#8467]
### Expected behavior
DBImplSecondary::TryCatchUpWithPrimary() should always catch up primary.

### Actual behavior
1. If we call TryCatchUpWithPrimary() in same thread before `secondary->Get()`, it is ok, newly added data(on primary) can be found.
2. If we keep calling TryCatchUpWithPrimary() in a dedicated thread, newly added data(on primary) can **NOT** be found.

### Steps to reproduce the behavior
write data on primary, and on secondary: keep calling TryCatchUpWithPrimary() in a dedicated thread and call `secondary->Get()` in another thread.

--------------------------------------------------------------------------

Current behavior of TryCatchUpWithPrimary is very less useful in real application.

**Comment by ajkr** [comment]
Just to confirm my understanding, are you saying the pattern always works?

1. T1 calls `TryCatchUpWithPrimary()`
2. T1 calls `secondary->Get()`

But moving step 2. to a separate thread, T2, while still ensuring step 1. completes first, breaks it? We would need to see a code example; a problem that is exposed in this way sounds surprising.

**Comment by rockeet** [comment]
Yes, here is our code snippets:

```c++
void CallTryCatchUpWithPrimary(rocksdb::DB* db) {
  while (true) {
    rocksdb::Status s = db->TryCatchUpWithPrimary(); // s.ok() is true and 
    std::cerr << "(thread) TryCatchUpWithPrimary() " << "s.ok() is " << s.ok() << std::endl;
    // this message is always printed, but secondary does not catch up in main thread
  }
}
int main(int argc, char* argv[]) {
  rocksdb::Options options;  options.create_if_missing = true;
  rocksdb::Status s = rocksdb::DB::OpenAsSecondary(options, "testdb", "secondary_path", &db);
  assert(s.ok());
  bool use_thread = atoi(argv[1]) != 0;
  if (use_thread) std::thread(CallTryCatchUpWithPrimary, db).detach();
  while (true) {
    std::string key, value;
    std::cin >> key;
    if (!use_thread) {
      s = db->TryCatchUpWithPrimary(); // this is ok, secondary can catch up with primary
      std::cerr << "(not thread) TryCatchUpWithPrimary() " << "s.ok() is " << s.ok() << std::endl;
    }
    s = db->Get(rocksdb::ReadOptions(), key, &value);
    if (s.ok())
      std::cout << value << std::endl;
    else
      std::cout << s.ToString() << std::endl;
  }
  delete db;
  return 0;
}
```

**Comment by rockeet** [comment]
> Just to confirm my understanding, are you saying the pattern always works?
> 
> 1. T1 calls `TryCatchUpWithPrimary()`
> 2. T1 calls `secondary->Get()`
> 
> But moving step 2. to a separate thread, T2, while still ensuring step 1. completes first, breaks it? We would need to see a code example; a problem that is exposed in this way sounds surprising.

Yes, your understand is right, here is our code snippets:

```c++
void CallTryCatchUpWithPrimary(rocksdb::DB* db) {
  while (true) {
    rocksdb::Status s = db->TryCatchUpWithPrimary(); // s.ok() is true and 
    std::cerr << "(thread) TryCatchUpWithPrimary() " << "s.ok() is " << s.ok() << std::endl;
    // this message is always printed, but secondary does not catch up in main thread
  }
}
int main(int argc, char* argv[]) {
  rocksdb::Options options;  options.create_if_missing = true;
  rocksdb::Status s = rocksdb::DB::OpenAsSecondary(options, "testdb", "secondary_path", &db);
  assert(s.ok());
  bool use_thread = atoi(argv[1]) != 0;
  if (use_thread) std::thread(CallTryCatchUpWithPrimary, db).detach();
  while (true) {
    std::string key, value;
    std::cin >> key;
    if (!use_thread) {
      s = db->TryCatchUpWithPrimary(); // this is ok, secondary can catch up with primary
      std::cerr << "(not thread) TryCatchUpWithPrimary() " << "s.ok() is " << s.ok() << std::endl;
    }
    s = db->Get(rocksdb::ReadOptions(), key, &value);
    if (s.ok())
      std::cout << value << std::endl;
    else
      std::cout << s.ToString() << std::endl;
  }
  delete db;
  return 0;
}
```

**Comment by ajkr** [comment]
Thanks, I tried it out and also found updates missing.

By putting a `sleep(5)` in the "(thread)" loop, I was able to see memtable updates after one `TryCatchUpWithPrimary()` but then disappear after a second `TryCatchUpWithPrimary()`. In the "(no thread)" case, I was also able to see a memtable update after one `TryCatchUpWithPrimary()` but not after a second. So my conclusion is `TryCatchUpWithPrimary()` works at first (regardless of what thread it happens in) but then may revert the updates later.

**Comment by riversand963** [comment]
I think I found the cause.
`TryCatchUpWithPrimary()` calls `ReadAndApply()` that uses `ManifestTailer`. The recent refactor of `ReadAndApply()` will cause `last_sequence_` to go backwards when `ManifestTailer` is in `kCatchUp` mode. Will send out a fix.

**Comment by stijnopheide** [comment]
Just for reference, I confirm that we see the same behavior in the Java client. First call to tryCatchUpWithPrimary from a different thread works, but the second one reverts the updates.

**Comment by riversand963** [comment]
I believe the fix mentioned in this issue is fixed in #8653. Closing this for now and feel free to reopen if needed.
